lib/functoria/impl.ml

```diff
@@ -42,16 +42,6 @@ and 'a device = ('a, abstract) Device.t
 
 let abstract t = Abstract t
 
-let rec app_has_no_arguments = function
-  | App { args = Cons _; _ } | Dev { args = Cons _; _ } -> false
-  | Dev { dev; _ } when String.equal (Device.module_name dev) "Unit" ->
-      (* special hack for Job.noop *)
-      false
-  | App _ | Dev _ -> true
-  | If { cond = _; branches; default } ->
-      app_has_no_arguments default
-      || List.exists (fun (_, branch) -> app_has_no_arguments branch) branches
-
```

Comment:
Why is this needed? Did the reason why this was added changed with that PR?

Reply:
From my comment above:

Another remark is about the "no argument check" -- I think this brief hack introduced by myself is not the right thing. We should ensure that main (and other jobs) have at least one argument -- but if we pass in one argument (of ()) for each "device" that doesn't need it, we'll add a lot of things for no good reason (and have again complexity, and it being hard to understand).

--> I think this "app_has_no_argument" check was a good first iteration, but what we actually want to check is that "start takes at least an argument" (to have it executed after a log reporter) -- but already with extra_deps and runtime_args we get arguments. So, indeed this removal is unrelated and should not be merged as is. It would be great to have a better check (esp. since e.g. `noop` is already excluded from the check).

Reply:
superseeded by #1544